File: c/optimal.c
#include "road_system.h"
#include "optimal.h"


void roadStep(Path * paths[2], Section *pSection) {
	int timeA = totalTime(paths[0]);
	int timeB = totalTime(paths[1]);
	int forwardTimeToA = timeA + pSection->a;
	int crossTimeToA = timeB + pSection->b + pSection->c;
	int forwardTimeToB = timeB + pSection->b;
	int crossTimeToB = timeA + pSection->a + pSection->c;

	Path * oldA = clonePath(paths[0]);

	if (forwardTimeToA <= crossTimeToA)
		addToPath(paths[0], newPoint(A, pSection->a));
	else {
		paths[0] = clonePath(paths[1]);
		addToPath(paths[0], newPoint(B, pSection->b));
		addToPath(paths[0], newPoint(C, pSection->c));
	}

	if (forwardTimeToB <= crossTimeToB)
		addToPath(paths[1], newPoint(B, pSection->b));
	else {
		paths[1] = clonePath(oldA);
		addToPath(paths[1], newPoint(A, pSection->a));
		addToPath(paths[1], newPoint(C, pSection->c));
	}
}

Path * calculateOptimalPath(System *pSystem) {
	Path * bestPaths[2];
	bestPaths[0] = newPath();
	bestPaths[1] = newPath();

	for (int i = 0; i < pSystem->numSections; i++) {
		roadStep(bestPaths, &(pSystem->sections[i]));
	}

	if (totalTime(bestPaths[0]) < totalTime(bestPaths[1]))
		return bestPaths[0];
	else
		return bestPaths[1];
}

File: rust/src/optimal.rs
use roads::{System, Path, Section, Point};

pub fn calculate(system: &System) -> Path {
    let mut paths : Vec<Path> = Vec::with_capacity(2);
    paths.push(Path::new());
    paths.push(Path::new());

	for section in &system.sections {
		road_step(&mut paths, section);
	}

	if paths[0].total_time < paths[1].total_time {
		paths[0].clone()
	} else {
		paths[1].clone()
	}
}

fn road_step(paths: &mut Vec<Path>, section: &Section) {
	let time_a = paths[0].total_time;
	let time_b = paths[1].total_time;
	let forward_time_to_a = time_a + section.distances[0];
	let cross_time_to_a = time_b + section.distances[1] + section.distances[2];
	let forward_time_to_b = time_b + section.distances[1];
	let cross_time_to_b = time_a + section.distances[0] + section.distances[2];

	let old_a = paths[0].clone();

	if forward_time_to_a <= cross_time_to_a {
		paths[0].add(Point::new("A".to_string(), section.distances[0]));
	} else {
		paths[0] = paths[1].clone();
		paths[0].add(Point::new("B".to_string(), section.distances[1]));
		paths[0].add(Point::new("C".to_string(), section.distances[2]));
	}

	if forward_time_to_b <= cross_time_to_b {
		paths[1].add(Point::new("B".to_string(), section.distances[1]));
	} else {
		paths[1] = old_a;
		paths[1].add(Point::new("A".to_string(), section.distances[0]));
		paths[1].add(Point::new("C".to_string(), section.distances[2]));
	}
}
